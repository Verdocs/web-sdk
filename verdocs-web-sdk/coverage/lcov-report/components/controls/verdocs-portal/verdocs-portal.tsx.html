
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for components/controls/verdocs-portal/verdocs-portal.tsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> / <a href="index.html">components/controls/verdocs-portal</a> verdocs-portal.tsx</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">66.66% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>30/45</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">35.48% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>11/31</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">76.92% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>10/13</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">66.66% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>28/42</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a>
<a name='L44'></a><a href='#L44'>44</a>
<a name='L45'></a><a href='#L45'>45</a>
<a name='L46'></a><a href='#L46'>46</a>
<a name='L47'></a><a href='#L47'>47</a>
<a name='L48'></a><a href='#L48'>48</a>
<a name='L49'></a><a href='#L49'>49</a>
<a name='L50'></a><a href='#L50'>50</a>
<a name='L51'></a><a href='#L51'>51</a>
<a name='L52'></a><a href='#L52'>52</a>
<a name='L53'></a><a href='#L53'>53</a>
<a name='L54'></a><a href='#L54'>54</a>
<a name='L55'></a><a href='#L55'>55</a>
<a name='L56'></a><a href='#L56'>56</a>
<a name='L57'></a><a href='#L57'>57</a>
<a name='L58'></a><a href='#L58'>58</a>
<a name='L59'></a><a href='#L59'>59</a>
<a name='L60'></a><a href='#L60'>60</a>
<a name='L61'></a><a href='#L61'>61</a>
<a name='L62'></a><a href='#L62'>62</a>
<a name='L63'></a><a href='#L63'>63</a>
<a name='L64'></a><a href='#L64'>64</a>
<a name='L65'></a><a href='#L65'>65</a>
<a name='L66'></a><a href='#L66'>66</a>
<a name='L67'></a><a href='#L67'>67</a>
<a name='L68'></a><a href='#L68'>68</a>
<a name='L69'></a><a href='#L69'>69</a>
<a name='L70'></a><a href='#L70'>70</a>
<a name='L71'></a><a href='#L71'>71</a>
<a name='L72'></a><a href='#L72'>72</a>
<a name='L73'></a><a href='#L73'>73</a>
<a name='L74'></a><a href='#L74'>74</a>
<a name='L75'></a><a href='#L75'>75</a>
<a name='L76'></a><a href='#L76'>76</a>
<a name='L77'></a><a href='#L77'>77</a>
<a name='L78'></a><a href='#L78'>78</a>
<a name='L79'></a><a href='#L79'>79</a>
<a name='L80'></a><a href='#L80'>80</a>
<a name='L81'></a><a href='#L81'>81</a>
<a name='L82'></a><a href='#L82'>82</a>
<a name='L83'></a><a href='#L83'>83</a>
<a name='L84'></a><a href='#L84'>84</a>
<a name='L85'></a><a href='#L85'>85</a>
<a name='L86'></a><a href='#L86'>86</a>
<a name='L87'></a><a href='#L87'>87</a>
<a name='L88'></a><a href='#L88'>88</a>
<a name='L89'></a><a href='#L89'>89</a>
<a name='L90'></a><a href='#L90'>90</a>
<a name='L91'></a><a href='#L91'>91</a>
<a name='L92'></a><a href='#L92'>92</a>
<a name='L93'></a><a href='#L93'>93</a>
<a name='L94'></a><a href='#L94'>94</a>
<a name='L95'></a><a href='#L95'>95</a>
<a name='L96'></a><a href='#L96'>96</a>
<a name='L97'></a><a href='#L97'>97</a>
<a name='L98'></a><a href='#L98'>98</a>
<a name='L99'></a><a href='#L99'>99</a>
<a name='L100'></a><a href='#L100'>100</a>
<a name='L101'></a><a href='#L101'>101</a>
<a name='L102'></a><a href='#L102'>102</a>
<a name='L103'></a><a href='#L103'>103</a>
<a name='L104'></a><a href='#L104'>104</a>
<a name='L105'></a><a href='#L105'>105</a>
<a name='L106'></a><a href='#L106'>106</a>
<a name='L107'></a><a href='#L107'>107</a>
<a name='L108'></a><a href='#L108'>108</a>
<a name='L109'></a><a href='#L109'>109</a>
<a name='L110'></a><a href='#L110'>110</a>
<a name='L111'></a><a href='#L111'>111</a>
<a name='L112'></a><a href='#L112'>112</a>
<a name='L113'></a><a href='#L113'>113</a>
<a name='L114'></a><a href='#L114'>114</a>
<a name='L115'></a><a href='#L115'>115</a>
<a name='L116'></a><a href='#L116'>116</a>
<a name='L117'></a><a href='#L117'>117</a>
<a name='L118'></a><a href='#L118'>118</a>
<a name='L119'></a><a href='#L119'>119</a>
<a name='L120'></a><a href='#L120'>120</a>
<a name='L121'></a><a href='#L121'>121</a>
<a name='L122'></a><a href='#L122'>122</a>
<a name='L123'></a><a href='#L123'>123</a>
<a name='L124'></a><a href='#L124'>124</a>
<a name='L125'></a><a href='#L125'>125</a>
<a name='L126'></a><a href='#L126'>126</a>
<a name='L127'></a><a href='#L127'>127</a>
<a name='L128'></a><a href='#L128'>128</a>
<a name='L129'></a><a href='#L129'>129</a>
<a name='L130'></a><a href='#L130'>130</a>
<a name='L131'></a><a href='#L131'>131</a>
<a name='L132'></a><a href='#L132'>132</a>
<a name='L133'></a><a href='#L133'>133</a>
<a name='L134'></a><a href='#L134'>134</a></td><td class="line-coverage quiet"><span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import {Component, EventEmitter, h, Host, Listen, Prop, Event} from '@stencil/core';
&nbsp;
const Z_INDEX = '10001';
&nbsp;
/**
 * Display a child component in a "portal", popping it out of the main DOM tree
 * to allow it to escape the bounds set by its parent.
 * @credit https://github.com/tomas-teston/stencil-portal for the basic
 * technique. This has been altered in a few ways to make it more friendly
 * to cases where there may be multiple portals on the page and provide more
 * alignment options for the child to be displayed.
 *
 * ```ts
 * &lt;div style="padding: 40px"&gt;
 *     &lt;div id="sample-tooltip" style="border: 1px solid green; padding: 3px 10px;"&gt;
 *       Tooltip Anchor
 *       &lt;verdocs-portal anchor="sample-tooltip" .align=${align} .voffset=${voffset}&gt;
 *         &lt;div style="border: 1px solid red; padding: 3px 10px;"&gt;Tooltip&lt;/div&gt;
 *       &lt;/verdocs-portal&gt;
 *     &lt;/div&gt;
 *   &lt;/div&gt;
 * ```
 */
@Component({
  tag: 'verdocs-portal',
  styleUrl: 'verdocs-portal.scss',
  shadow: false,
})
export class VerdocsPortal {
  private portal: HTMLElement;
  private element: HTMLElement;
  private moved: boolean = false;
&nbsp;
  /**
   * Unique ID of the parent element to anchor to.
   */
  @Prop() anchor: string;
&nbsp;
  /**
   * Vertical offset from the parent.
   */
  @Prop() voffset: number = 0;
&nbsp;
  @Event() clickAway: EventEmitter&lt;void&gt;;
&nbsp;
  @Listen('scroll', {target: 'window', capture: true})
<span class="fstat-no" title="function not covered" >  handleScroll(</span>) {
<span class="cstat-no" title="statement not covered" >    this.calculatePosition();</span>
  }
&nbsp;
  @Listen('resize', {target: 'window'})
<span class="fstat-no" title="function not covered" >  handleResize(</span>) {
<span class="cstat-no" title="statement not covered" >    this.calculatePosition();</span>
  }
&nbsp;
  @Listen('click', {target: 'document'})
  handleClick(e) {
    const closest = e.target.closest('.verdocs-portal');
    if (!this.element.contains(e.target) &amp;&amp; !closest) {
      this.clickAway<span class="branch-0 cbranch-no" title="branch not covered" >?.e</span>mit();
    }
&nbsp;
    // Temporary hack to dismiss portals-in-portals, which right now is pretty much just
    // multiselect.
    document.querySelectorAll('.verdocs-multiselect-dropdown').forEach(<span class="fstat-no" title="function not covered" >el </span>=&gt; <span class="cstat-no" title="statement not covered" >el.remove())</span>;
  }
&nbsp;
  private calculateLeft() {
    const anchorEl = document.getElementById(this.anchor);
    if (!anchorEl) return 0;
&nbsp;
    const anchorRect = <span class="cstat-no" title="statement not covered" >anchorEl.getBoundingClientRect();</span>
&nbsp;
    const rightOfViewportWithScroll = <span class="cstat-no" title="statement not covered" >document.documentElement.clientWidth + document.documentElement.scrollLeft;</span>
    const leftOfPopup = <span class="cstat-no" title="statement not covered" >Math.max(anchorRect.left, 0);</span>
    // Old centering logic, disabled temporarily
    //    Math.max(anchorRect.left - this.portal.offsetWidth / 2 + anchorRect.width / 2, 0);
    // Old right-align logic, disabled temporarily
    //    Math.max(anchorRect.left + anchorRect.width - this.portal.offsetWidth, 0);
    const popupWidth = <span class="cstat-no" title="statement not covered" >this.element?.offsetWidth || 400;</span>
    const offRightEdge = <span class="cstat-no" title="statement not covered" >leftOfPopup + popupWidth &gt; rightOfViewportWithScroll;</span>
&nbsp;
<span class="cstat-no" title="statement not covered" >    return offRightEdge ? rightOfViewportWithScroll - popupWidth - 20 : leftOfPopup;</span>
  }
&nbsp;
  private calculateTop() {
    const anchorEl = document.getElementById(this.anchor);
    if (!anchorEl) return 0;
&nbsp;
    const anchorRect = <span class="cstat-no" title="statement not covered" >anchorEl.getBoundingClientRect();</span>
&nbsp;
    const bottomOfViewportWithScroll = <span class="cstat-no" title="statement not covered" >document.documentElement.clientHeight + document.documentElement.scrollTop;</span>
    const topOfPopup = <span class="cstat-no" title="statement not covered" >anchorRect.bottom + this.voffset + document.documentElement.scrollTop;</span>
    const popupHeight = <span class="cstat-no" title="statement not covered" >this.element?.offsetHeight || 300;</span>
    const offBottomEdge = <span class="cstat-no" title="statement not covered" >topOfPopup + popupHeight &gt; bottomOfViewportWithScroll;</span>
&nbsp;
<span class="cstat-no" title="statement not covered" >    return offBottomEdge ? topOfPopup - popupHeight - anchorRect.height : topOfPopup;</span>
  }
&nbsp;
  private calculatePosition() {
    this.portal.style.top = `${this.calculateTop()}px`;
    this.portal.style.left = `${this.calculateLeft()}px`;
  }
&nbsp;
  componentWillLoad() {
    const id = `${this.anchor}-portal`;
    document.getElementById(id)<span class="branch-1 cbranch-no" title="branch not covered" ><span class="branch-1 cbranch-no" title="branch not covered" >?.r</span>emove();</span>
&nbsp;
    this.portal = document.createElement('div');
    this.portal.setAttribute('id', id);
    this.portal.style.zIndex = Z_INDEX;
    this.portal.style.position = 'absolute';
    this.portal.className = 'verdocs-portal-wrapper';
    document.body.append(this.portal);
  }
&nbsp;
  componentDidLoad() {
    this.portal.appendChild(this.element);
    this.calculatePosition();
  }
&nbsp;
  disconnectedCallback() {
    this.moved ? <span class="branch-0 cbranch-no" title="branch not covered" >this.portal.remove() </span>: (this.moved = true);
  }
&nbsp;
  render() {
    return (
      &lt;Host ref={el =&gt; (this.element = el)} class="verdocs-portal"&gt;
        &lt;slot /&gt;
      &lt;/Host&gt;
    );
  }
}
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-07-24T13:19:16.585Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    