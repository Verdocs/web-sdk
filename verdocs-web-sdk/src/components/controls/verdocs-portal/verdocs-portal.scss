@import '../../../theme.scss';

verdocs-portal {
}
