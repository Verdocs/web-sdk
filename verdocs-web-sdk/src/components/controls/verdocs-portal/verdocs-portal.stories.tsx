import {html} from 'lit-html';
import {Meta} from '@storybook/web-components';

export default {
  title: 'Controls/Portal',
  component: 'verdocs-portal',
  args: {
    align: 'left',
    voffset: 0,
  },
  argTypes: {},
} as Meta;

export const Portal = ({align, voffset}) => html`
  <div style="padding: 40px">
    <div id="sample-tooltip" style="border: 1px solid green; padding: 3px 10px;">
      Tooltip Anchor
      <verdocs-portal anchor="sample-tooltip" .align=${align} .voffset=${voffset}>
        <div style="border: 1px solid red; padding: 3px 10px;">Tooltip</div>
      </verdocs-portal>
    </div>
  </div>
`;
