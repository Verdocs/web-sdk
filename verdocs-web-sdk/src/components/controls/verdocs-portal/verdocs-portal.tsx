import {Component, EventEmitter, h, Host, Listen, Prop, Event} from '@stencil/core';

const Z_INDEX = '10001';

/**
 * Display a child component in a "portal", popping it out of the main DOM tree
 * to allow it to escape the bounds set by its parent.
 * @credit https://github.com/tomas-teston/stencil-portal for the basic
 * technique. This has been altered in a few ways to make it more friendly
 * to cases where there may be multiple portals on the page and provide more
 * alignment options for the child to be displayed.
 *
 * ```ts
 * <div style="padding: 40px">
 *     <div id="sample-tooltip" style="border: 1px solid green; padding: 3px 10px;">
 *       Tooltip Anchor
 *       <verdocs-portal anchor="sample-tooltip" .align=${align} .voffset=${voffset}>
 *         <div style="border: 1px solid red; padding: 3px 10px;">Tooltip</div>
 *       </verdocs-portal>
 *     </div>
 *   </div>
 * ```
 */
@Component({
  tag: 'verdocs-portal',
  styleUrl: 'verdocs-portal.scss',
  shadow: false,
})
export class VerdocsPortal {
  private portal: HTMLElement;
  private element: HTMLElement;
  private moved: boolean = false;

  /**
   * Unique ID of the parent element to anchor to.
   */
  @Prop() anchor: string;

  /**
   * Vertical offset from the parent.
   */
  @Prop() voffset: number = 0;

  @Event() clickAway: EventEmitter<void>;

  @Listen('scroll', {target: 'window', capture: true})
  handleScroll() {
    this.calculatePosition();
  }

  @Listen('resize', {target: 'window'})
  handleResize() {
    this.calculatePosition();
  }

  @Listen('click', {target: 'document'})
  handleClick(e) {
    if (!this.element.contains(e.target)) {
      this.clickAway?.emit();
    }
  }

  private calculateLeft() {
    const anchorEl = document.getElementById(this.anchor);
    if (!anchorEl) return 0;

    const anchorRect = anchorEl.getBoundingClientRect();

    const rightOfViewportWithScroll = document.documentElement.clientWidth + document.documentElement.scrollLeft;
    const leftOfPopup = Math.max(anchorRect.left, 0);
    // Old centering logic, disabled temporarily
    //    Math.max(anchorRect.left - this.portal.offsetWidth / 2 + anchorRect.width / 2, 0);
    // Old right-align logic, disabled temporarily
    //    Math.max(anchorRect.left + anchorRect.width - this.portal.offsetWidth, 0);
    const popupWidth = this.element?.offsetWidth || 400;
    const offRightEdge = leftOfPopup + popupWidth > rightOfViewportWithScroll;

    return offRightEdge ? leftOfPopup + anchorRect.width - popupWidth : leftOfPopup;
  }

  private calculateTop() {
    const anchorEl = document.getElementById(this.anchor);
    if (!anchorEl) return 0;

    const anchorRect = anchorEl.getBoundingClientRect();

    const bottomOfViewportWithScroll = document.documentElement.clientHeight + document.documentElement.scrollTop;
    const topOfPopup = anchorRect.bottom + this.voffset + document.documentElement.scrollTop;
    const popupHeight = this.element?.offsetHeight || 300;
    const offBottomEdge = topOfPopup + popupHeight > bottomOfViewportWithScroll;

    return offBottomEdge ? topOfPopup - popupHeight - anchorRect.height : topOfPopup;
  }

  private calculatePosition() {
    this.portal.style.top = `${this.calculateTop()}px`;
    this.portal.style.left = `${this.calculateLeft()}px`;
  }

  componentWillLoad() {
    const id = `${this.anchor}-portal`;
    document.getElementById(id)?.remove();

    this.portal = document.createElement('div');
    this.portal.setAttribute('id', id);
    this.portal.style.zIndex = Z_INDEX;
    this.portal.style.position = 'absolute';
    document.body.append(this.portal);
  }

  componentDidLoad() {
    this.portal.appendChild(this.element);
    this.calculatePosition();
  }

  disconnectedCallback() {
    this.moved ? this.portal.remove() : (this.moved = true);
  }

  render() {
    return (
      <Host ref={el => (this.element = el)}>
        <slot />
      </Host>
    );
  }
}
