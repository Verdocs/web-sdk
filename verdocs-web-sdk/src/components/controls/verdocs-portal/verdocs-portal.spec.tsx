import { newSpecPage } from '@stencil/core/testing';
import { VerdocsPortal } from './verdocs-portal';

describe('verdocs-portal', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsPortal],
      html: `<verdocs-portal></verdocs-portal>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with anchor and voffset', async () => {
    const page = await newSpecPage({
      components: [VerdocsPortal],
      html: `<verdocs-portal anchor="anchor-id" voffset="10"></verdocs-portal>`,
    });
    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.anchor).toBe('anchor-id');
    expect(page.rootInstance.voffset).toBe(10);
  });

  it('emits clickAway event on outside click', async () => {
    const page = await newSpecPage({
      components: [VerdocsPortal],
      html: `<verdocs-portal></verdocs-portal>`,
    });
    const spy = jest.fn();
    page.root.addEventListener('clickAway', spy);

    // Simulate a click outside the portal
    (page.rootInstance as any).handleClick({ target: document.createElement('div') });
    expect(spy).toHaveBeenCalled();
  });
});
