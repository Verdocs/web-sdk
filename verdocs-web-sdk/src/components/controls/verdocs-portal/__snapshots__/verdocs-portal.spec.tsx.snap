// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-portal renders with anchor and voffset 1`] = `
<verdocs-portal anchor="anchor-id" class="verdocs-portal" voffset="10">
  <!---->
</verdocs-portal>
`;

exports[`verdocs-portal renders with default props 1`] = `
<verdocs-portal class="verdocs-portal">
  <!---->
</verdocs-portal>
`;
